Adding annotation to drive the name of the backing topic

matzew opened the pull request:
Signed-off-by: Matthias Wessendorf <mwessend@redhat.com>

Fixes #977 

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- adding support for `kafka.eventing.knative.dev/external.topic` annotation for external topic usage

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
With the `kafka.eventing.knative.dev/external.topic` annotation it is possible to use an externally managed Apache Kafka topic for the broker usage
```

**Docs**

<!--
:book: If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

slinkydeveloper commented:
Nice to see this!

I suggest this PR also includes:

* a bunch of unit tests
* an e2e test (just check the tests in `e2e_new`)

We will also need a PR to the docs repo for that (I would recommend specifying that the feature is still unstable)

matzew commented:
yeah, is not ready yet

On Mon 21. Jun 2021 at 17:02, Francesco Guardiani ***@***.***>
wrote:

> Nice to see this!
>
> I suggest this PR also includes:
>
>    - a bunch of unit tests
>    - an e2e test (just check the tests in e2e_new)
>
> We will also need a PR to the docs repo for that (I would recommend
> specifying that the feature is still unstable)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative-sandbox/eventing-kafka-broker/pull/1023#issuecomment-865106220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABGPTS7A2IPPQBY5OCQVA3TT5IAXANCNFSM47BYFQ7A>
> .
>
-- 
Sent from Gmail Mobile

slinkydeveloper commented on control-plane/pkg/reconciler/broker/broker.go:
```diff
@@ -212,6 +213,20 @@ func (r *Reconciler) reconcileKind(ctx context.Context, broker *eventing.Broker)
+}
+
+func isCustomTopic(broker *eventing.Broker) (string, bool) {
+	topicAnnotationValue, ok := broker.Annotations[topicAnnotation]
```
Maybe just return `bool` since the function name starts with `is`?

matzew replied:
yeah, this is not ready. I am just adding to my WIP

amir-hadi commented:
Hi folks, is there anything we could do to get this into the release? We will need this functionality :)

matzew commented:
@amir-hadi thanks for the feedback. I just returned from PTO, and will see to finish this PR soon.

amir-hadi commented:
Hey matzew, this wasn't at all to pressure you. I am just thinking if I can spare some engineering time to get this done and help you, as this will help us to use the best of both worlds: Stream processing with kSQL, topic management by strimzi and connecting KNative services with events using the broker implementation of yours.

matzew commented:
@amir-hadi  yeah, sounds good

I think we need to be a bit mindful here, if people bring their own topic - we need to be safe that they don't write any bogus (for what ever reason) kafka records to the topic

maganaluis commented:
Hi @matzew, I'm also a little eager for this feature as well, we are currently using Azure Event Hub which has a sad limit of only 10 topics, I just tested your code with that configuration and works great! I'd love to see it in the next release, thank you for all your work!

dahateb commented:
Hey all, we would like to have this functionality in to be able to use our current topics with the broker

matzew commented:
Hi @maganaluis  and @dahateb 

I think exposing the topic name is not that simple. We have to be careful, since that is an implementation detail of the broker implementation. E.g. this does not work if non-CloudEvent records are stored in the topic.

I think we need to fiddle a bit more on the design, rather than keeping it as my PR proposes it .   

I will need to think about this a bit more :disappointed:

matzew commented:
I was wondering if we could use some `Kafka Connect` connector here, that would read from Kafka topic (as the source).

The Connector itself would have a SMT (see: https://cwiki.apache.org/confluence/display/KAFKA/KIP-66%3A+Single+Message+Transforms+for+Kafka+Connect)  that could than rewrite the payload as CloudEvent, to that topic that is backing the broker....

amir-hadi commented:
@matzew just to add a bit more context. We want to have CloudEvent messages, but we have a topic naming schema that doesn't work with the current implementation. We have a naming schema that includes datacenter, region etc. which is valuable information for replication purposes.

Hence I would be grateful to have your PR already :)

maganaluis commented:
@matzew How about discarding the message if is not a well formatted CE? For instance, if you have a DeadLetterSink which responds with 'success' rather than a CE, though this is more on the Pub side but I did encounter this. 

I think there are two alternatives, loosen the CE restrictions, or enforce them more rigorously. My thinking is that if those messages are not CE, they were not meant to be processed anyway.

dahateb commented:
any progress here?

matzew commented:
Hey folks - thanks for the interest. I will rethink this and see what's the best to get this landed .

matzew commented:
Without this PR - would it be possible to cover your needs by using a combination of the `KafkaSource` (with a specific topic, on a specific kafka-broker), and have that emit the Kafka Records as CloudEvents to the Knative-Kafka-Broker?

jhill072 commented:
Just to throw a couple more pennies to the pot, we have an enforced naming convention for topics, which means we cannot create the topic with the name predetermined by this broker. Further, there is another requirement of the topics being created through Argo (and not programmatically), which prevents the reconciler from creating the topic. We would still like the broker to manage everything in the topic, but we just need a different name than the one being used currently.

So I would still like to vote for this PR to merge, pretty please!

matzew commented:
I've updated the PR and changed it slightly.

I am using an `x-` annotation, and a note that this is a none-public feature, and this can go away in future releases.

I do understand the comments in the ticket, that people do want this feature...

matzew commented:
Perhaps until we have a better way to "import" from Kafka topics, this might be a temporary thing?

With the `WARNING` to not really use it - as it will potentially go away? 


Question: Are folks currently patching their code-base to be able to do this?

jhill072 commented:
> Question: Are folks currently patching their code-base to be able to do this?

That is currently my plan of action if push comes to shove with my company's gatekeepers

devguyio commented:
@matzew @jhill072 thanks for picking up the discussion. One of the reasons this was not right away merged is that one of the design principles of the Broker is to be a black box, how it manages its topics, the topics names etc is an internal implementation detail that can change with time. 

And the direction we're heading towards is to allow a smarter `Source + Broker` that don't go through the unnecessary kafka -> http -> kafka round trip.

That one is going to be tracked here #1580 .

> Just to throw a couple more pennies to the pot, we have an enforced naming convention for topics, which means we cannot create the topic with the name predetermined by this broker. Further, there is another requirement of the topics being created through Argo (and not programmatically), which prevents the reconciler from creating the topic. We would still like the broker to manage everything in the topic, but we just need a different name than the one being used currently.

However, this use case described above probably means that you can't use the Kafka Broker without the ability to provide the topic details. So with this I'm having a change of heart in maybe considering adding this as an [experimental feature](https://knative.dev/development/eventing/experimental-features/)? while at the same time discussing the following use case in depth. What does it mean to have a broker that doesn't manage its topics?

I wonder if that would be considered a good solution?

jhill072 commented:
Alternatively, to avoid the situation of describing the generic Broker with
implementation-specific annotations, we could extend the config map to
include the topic name, similar to the other topic details?

We've recently ran into another situation that requires setting retention
period differently but since the config map only supports partitions and
replication, we'll have to make the topic elsewhere.

Thoughts on adding to the config map approach?

On Tue, Dec 14, 2021, 1:38 PM Matthias Wessendorf ***@***.***>
wrote:

> yep, I can update the PR to go the extra mile of being experimental
> feature - not for 1.1
>
> but for 1.2
>
> I think this is good, moving forward
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative-sandbox/eventing-kafka-broker/pull/1023#issuecomment-993869493>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG7245BGY4BIGOFVGAYKAM3UQ6FMBANCNFSM47BYFQ7A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

matzew commented:
I dont think the confiap is a good idea

I will do the experimental feature

On Wed 15. Dec 2021 at 13:53, jhill072 ***@***.***> wrote:

> Alternatively, to avoid the situation of describing the generic Broker with
> implementation-specific annotations, we could extend the config map to
> include the topic name, similar to the other topic details?
>
> We've recently ran into another situation that requires setting retention
> period differently but since the config map only supports partitions and
> replication, we'll have to make the topic elsewhere.
>
> Thoughts on adding to the config map approach?
>
> On Tue, Dec 14, 2021, 1:38 PM Matthias Wessendorf ***@***.***>
> wrote:
>
> > yep, I can update the PR to go the extra mile of being experimental
> > feature - not for 1.1
> >
> > but for 1.2
> >
> > I think this is good, moving forward
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/knative-sandbox/eventing-kafka-broker/pull/1023#issuecomment-993869493
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AG7245BGY4BIGOFVGAYKAM3UQ6FMBANCNFSM47BYFQ7A
> >
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <
> https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675
> >
> > or Android
> > <
> https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub
> >.
> >
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative-sandbox/eventing-kafka-broker/pull/1023#issuecomment-994763143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABGPTUDVCCTX24YFENEUMLURCFSTANCNFSM47BYFQ7A>
> .
>
-- 
Sent from Gmail Mobile

pierDipi commented:
> We've recently ran into another situation that requires setting retention
period differently but since the config map only supports partitions and
replication, we'll have to make the topic elsewhere.

Can you please open an issue for the retention period?

jhill072 commented:
> > We've recently ran into another situation that requires setting retention
> > period differently but since the config map only supports partitions and
> > replication, we'll have to make the topic elsewhere.
> 
> Can you please open an issue for the retention period?

As requested! https://github.com/knative-sandbox/eventing-kafka-broker/issues/1638
Let me know if I can help out with this at all

pierDipi reviewed (commented):
Can we add e2e test?

matzew commented on control-plane/pkg/reconciler/broker/broker.go:
```diff
@@ -48,6 +48,10 @@ import (
+
+	// private annotation for changing the topic
+	// NOTE: this may go away in a future release
+	externalTopicAnnotation = "x-kafka.eventing.knative.dev/external.topic"
```
one question I have, @pierDipi 
I think... we can also name this `kafka.eventing.knative.dev/external.topic` and skip the `x-`

I do like this as a proper feature, tbh

pierDipi replied:
+1

matzew commented:
/retest

matzew commented on test/e2e/broker_trigger_test.go:
```diff
@@ -268,3 +269,141 @@ func TestBrokerWithConfig(t *testing.T) {
+
+// WithExternalTopicForBroker returns a function that adds an
+// annotation to the given v1 (Kafka) Broker for using an external managed topic
+func WithExternalTopicForBroker(annotationValue string) BrokerOption {
```
I think I will locate this somewhere else - but there is no real of a `lib`

That said, I think generally handy for upstream eventing is something like:

```
func WithCustomAnnotationForBroker(annotationKey, annotationValue string) BrokerOption {...}
```

Which _our_ `func` here could make use of ...

matzew commented:
/hold

once the dependencies are updated, I will remove the custom func in the code

matzew commented:
/retest

pierDipi commented on test/e2e/broker_trigger_test.go:
```diff
@@ -268,3 +270,110 @@ func TestBrokerWithConfig(t *testing.T) {
 	})
 }
+
+func TestBrokerExternalTopicTrigger(t *testing.T) {
+	testingpkg.RunMultiple(t, func(t *testing.T) {
```
Can we use the new framework for new tests?

pierDipi replied:
In addition, this test doesn't prevent regression in the feature since I can just drop the new reconciler logic and it will still work, what about using a custom role and the permissions to only publish and subscribe to a particular topic?

pierDipi replied:
> > Can we use the new framework for new tests?
> 
> you mean the reconciler-test? I decided against that since I'd have to extend/customize the broker, since the upstream template doe not really allow (n) custom annotations.

we can always create our own Kafka broker template

matzew commented:
/test unit-tests_eventing-kafka-broker_main

matzew commented:
/unhold

@pierDipi  can you take a look, I added a test for the reconciler

NEXT: I will port my prev e2e test to the reconciler test framework on a different PR, but would be good to land this feature now on the next upstream release

pierDipi commented on control-plane/pkg/reconciler/broker/broker_test.go:
```diff
@@ -2085,6 +2189,22 @@ func useTable(t *testing.T, table TableTest, env *config.Env) {
 		}
 
+		externalTopicUsage := false
+		expectedTopicName := fmt.Sprintf("%s%s-%s", TopicPrefix, BrokerNamespace, BrokerName)
+		if _, ok := row.OtherTestData[externalTopic]; ok {
+			expectedTopicName = ExternalTopicName
+			externalTopicUsage = true
+		}
```
I would use the value associated with `OtherTestData[externalTopic]` to retrieve the name instead of using the global constant here.

pierDipi replied:
```suggestion
		expectedTopicName := fmt.Sprintf("%s%s-%s", TopicPrefix, BrokerNamespace, BrokerName)
		if t, ok := row.OtherTestData[externalTopic]; ok {
			expectedTopicName = t
			externalTopicUsage = true
		}
```

matzew replied:
ok, true used now `expectedTopicName = t.(string)`

pierDipi commented on control-plane/pkg/reconciler/broker/broker_test.go:
```diff
@@ -2105,11 +2225,13 @@ func useTable(t *testing.T, table TableTest, env *config.Env) {
+					ExpectedTopicDetail:                    expectedTopicDetail,
+					ErrorOnCreateTopic:                     onCreateTopicError,
+					ErrorOnDeleteTopic:                     onDeleteTopicError,
+					ExpectedTopics:                         []string{ExternalTopicName},
```
why is this always set to `ExternalTopicName`?

pierDipi commented on control-plane/pkg/reconciler/broker/broker_test.go:
```diff
@@ -2085,6 +2189,22 @@ func useTable(t *testing.T, table TableTest, env *config.Env) {
+			externalTopicUsage = true
+		}
+
+		var metadata []*sarama.TopicMetadata
+		if externalTopicUsage {
+			metadata = append(metadata, &sarama.TopicMetadata{
+				Name:       ExternalTopicName,
+				IsInternal: false,
+				Partitions: []*sarama.PartitionMetadata{{}},
+			})
+		}
```
why is this special case needed?
It seems that these metadata are good in both cases.

matzew commented:
@pierDipi updated PR

pierDipi commented on control-plane/pkg/reconciler/testing/objects_common.go:
```diff
@@ -374,6 +374,15 @@ func StatusFailedToCreateTopic(topicName string) func(obj duckv1.KRShaped) {
 		)
 	}
 }
+func StatusTopicNotPresentOrInvalid(topicName string) func(obj duckv1.KRShaped) {
```
```suggestion

func StatusTopicNotPresentOrInvalid(topicName string) func(obj duckv1.KRShaped) {
```

pierDipi replied:
nit: empty line between the 2 functions, I guess the linter will complain otherwise

pierDipi commented:
@matzew do we have docs already / planned (any open issue?) ?

matzew commented:
/test unit-tests_eventing-kafka-broker_main

matzew commented:
@pierDipi I created a PR for the docs: https://github.com/knative/docs/pull/4981

pierDipi reviewed (approved):
/lgtm
/approve

matzew commented:
/cherry-pick release-1.4

matzew commented:
With 1.5 we have this now in. and see the documentation:
* https://knative.dev/development/eventing/broker/kafka-broker/#bring-your-own-topic

/cc @maganaluis , @amir-hadi, @dahateb
